Fix `ArrayBufferView`, `ArrayOps.ArrayIterator` and `View.Patched`

NthPortal opened the pull request:
Ensure `ArrayBufferView` is consistent with its buffer.

Simplify `ArrayBuffer#insertAll` code.

Fix unreported bug in `ArrayOps.ArrayIterator` where
index in array can overflow to negative, and in such cases
the iterator reports `hasNext` incorrectly as `true`.

Improve code documentation and readability of
`Iterator.patch` implementation, which was bumped
into while investigating the bug.

Discovered while writing tests for bug in `View.Patched`
(see below).

Fix unreported bug in `View.Patched` where iterator
is incorrect due to the patch only being iterable once,
and already having been exhausted.

Discovered while attempting to optimise `ArrayBuffer` alongside
fixing `ArrayBufferView` (eventually scrapped).

Fixes scala/bug#12284

SethTisue commented:
/sync

NthPortal commented:
/sync

NthPortal commented:
Note: 995f0aa and 82845aa did build fine, but for some reason their jobs have vanished

SethTisue commented:
/rebuild 82845a

SethTisue commented:
/rebuild 995f0a

SethTisue commented:
Rebasing would probably un-confuse Jenkins, here.

NthPortal commented:
I'll give it a try

julienrf commented on src/library/scala/collection/View.scala:
```diff
@@ -404,9 +404,15 @@ object View extends IterableFactory[View] {
   @SerialVersionUID(3L)
   private[collection] class Patched[A](underlying: SomeIterableOps[A], from: Int, other: IterableOnce[A], replaced: Int) extends AbstractView[A] {
-    def iterator: Iterator[A] = underlying.iterator.patch(from, other.iterator, replaced)
+    // we may be unable to traverse `other` more than once, so we need to cache it if that's the case
```
In which situation may we traverse `other` more than once?

NthPortal replied:
in principle, any view can be traversed arbitrarily many times, so `coll.view.patched(1, Iterator.single(1), 1)` could produce an incorrect result without this change if traversed twice (which is permitted)

NthPortal replied:
I think what caused this issue to crop up for me was the change to call `size` on `Iterable`s in `ArrayBuffer#addAll`

NthPortal replied:
incidentally, you could already hit this bug without going out of your way using `insertAll`:

```scala
scala> ArrayBuffer(1)
     |   .insertAll(
     |     1,
     |     List(2)
     |       .view
     |       .patch(1, List(3, 4, 5).iterator, 0)
     |   )
java.util.NoSuchElementException: head of empty list
  at scala.collection.immutable.Nil$.head(List.scala:629)
  at scala.collection.immutable.Nil$.head(List.scala:628)
  at scala.collection.StrictOptimizedLinearSeqOps$$anon$1.next(LinearSeq.scala:253)
  at scala.collection.Iterator$$anon$17.next(Iterator.scala:915)
  at scala.collection.mutable.ArrayBuffer.insertAll(ArrayBuffer.scala:181)
  ... 32 elided

scala> ArrayBuffer(1)
     |   .insertAll(
     |     1,
     |     List(2)
     |       .view
     |       .patch(1, List(3, 4, 5), 0)
     |   )

scala>
```

lrytz replied:
could be added as a test case to this PR?

NthPortal replied:
sure

NthPortal replied:
there's a simpler repro that we can use as a test, even.

```scala
scala> val view = List(2).view.patch(1, List(3, 4, 5).iterator, 0)
val view: scala.collection.View[Int] = View(<not computed>)

scala> view.toList
val res0: List[Int] = List(2, 3, 4, 5)

scala> view.toList
val res1: List[Int] = List(2)
```

julienrf commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -295,8 +309,48 @@ object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
+// TODO: use `CheckedIndexedSeqView.Id` once we can change the return type of
+final class ArrayBufferView[A] private[mutable](underlying: ArrayBuffer[A], mutationCount: () => Int)
+  extends AbstractIndexedSeqView[A] {
+  @deprecated("never intended to be public; call ArrayBuffer#view instead", since = "2.13.4")
```
```suggestion
  @deprecated("never intended to be public; call ArrayBuffer#view instead", since = "2.13.5")
```

NthPortal replied:
dah, it's 2.13.6 now

julienrf commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -295,8 +309,48 @@ object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
 
-final class ArrayBufferView[A](val array: Array[AnyRef], val length: Int) extends AbstractIndexedSeqView[A] {
+// TODO: use `CheckedIndexedSeqView.Id` once we can change the return type of
+final class ArrayBufferView[A] private[mutable](underlying: ArrayBuffer[A], mutationCount: () => Int)
```
What is the benefit of passing `mutationCount` as a separate parameter, rather than making it `private[mutable]` in class `ArrayBuffer`?

NthPortal replied:
none that I can think of. I think the reason I did it this way was just to mirror the regular `CheckedIndexedSeqView`, which is what we'll hopefully use eventually. I don't really have strong preferences

SethTisue replied:
I'm going to hit "resolve" here since nobody has expressed a strong opinion either way.

julienrf commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -135,10 +146,15 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
-        ensureSize(length + elems.length)
-        Array.copy(elems.array, 0, array, length, elems.length)
-        size0 = length + elems.length
+      case elems: collection.Iterable[A] =>
```
Do we want to match `IndexdSeq` only, since we call `elems.size` below?

NthPortal replied:
I'm just copying what `insertAll` already does.

I think the rationale is that it's cheaper to iterate over the collection twice than to allocate a whole bunch of (increasingly large) arrays repeatedly.

side note: both this method and `insertAll` probably ought to have a case handling `IterableOnce` where `knownSize >= 0` as well

NthPortal replied:
> I think the rationale is that it's cheaper to iterate over the collection twice than to allocate a whole bunch of (increasingly large) arrays repeatedly.

the benchmarks do not appear to agree with this

NthPortal replied:
I ended up checking `knownSize`, which is slightly broader than `IndexedSeq`, but still cheap

julienrf commented on project/MimaFilters.scala:
```diff
@@ -24,6 +24,26 @@ object MimaFilters extends AutoPlugin {
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.convert.JavaCollectionWrappers#IteratorWrapper.asIterator"),
+
+    // Fixes for scala/bug#12009
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.mutable.ArrayBufferView.this"),
```
IIUC it’s OK to add these filters because this method and the following ones are package private, right?

NthPortal replied:
yes. I forgot to note that in a comment - I'll try to do that

NthPortal replied:
fixed in the other PR's branch - will rebase on that once merged

julienrf reviewed (approved):
Thank you so much @NthPortal! (for your patience also…!)

NthPortal commented:
@julienrf it's probably a good idea to merge #9258 (which this is based on) first

SethTisue commented:
(marking as draft until #9258 lands)

lrytz commented:
LGTM, again leaving to @SethTisue

SethTisue commented:
I have started working on this, but it's not going to make 2.13.6. I'll make it a priority for early in the 2.13.7 process.

SethTisue commented:
@NthPortal hard to be totally sure, but afaict the `ArrayBufferBenchmark` I added over in #9258 should suffice for benchmarking this one as well

NthPortal commented:
@SethTisue yes.

I'll have to review the changes, as it's been a really long time since I've looked at them, but I'm not even sure there's anything to benchmark

NthPortal commented on test/junit/scala/collection/ViewTest.scala:
```diff
@@ -113,4 +113,11 @@ class ViewTest {
+
+  // see https://github.com/scala/scala/pull/9388/files#r630681050
+  @Test
+  def patch(): Unit = {
```
dah, this test is in the wrong commit

SethTisue commented:
needs rebase on top of #9258, which is now merged

NthPortal committed:
Add more addAll benchmarks for `ArrayBuffer`

NthPortal committed:
[bug#12284] Ensure `ArrayBufferView` is consistent with its buffer

Ensure `ArrayBufferView` is consistent with its buffer.

Simplify `ArrayBuffer#insertAll` code.

NthPortal commented:
I'm benching `addAll` after changes at the moment. thankfully, running fewer benchmarks takes less time

NthPortal commented:
I've written an Ammonite script to merge before and after JMH output files as a diff, if anyone wants it: https://gist.github.com/NthPortal/b21ea1d91f41a1c82a3113c0c3848da4

features:

- strip output prefix (usually `[info] ` in sbt)
- align the columns of the two outputs (most frequently misaligned in the "Error" column)
- deduplicates header when interleaving

NthPortal commented:
benchmark results

```diff
-before
+after

 Benchmark                     (size)  Mode  Cnt      Score     Error  Units
-ArrayBufferBenchmark.addAll1      10  avgt   30    104.579 ±   3.018  ns/op
+ArrayBufferBenchmark.addAll1      10  avgt   30     99.248 ±   0.430  ns/op
-ArrayBufferBenchmark.addAll1     100  avgt   30    266.944 ±   0.580  ns/op
+ArrayBufferBenchmark.addAll1     100  avgt   30    266.565 ±   0.470  ns/op
-ArrayBufferBenchmark.addAll1    1000  avgt   30   1699.199 ±   7.797  ns/op
+ArrayBufferBenchmark.addAll1    1000  avgt   30   1697.700 ±   4.453  ns/op
-ArrayBufferBenchmark.addAll1   10000  avgt   30  24915.264 ±  50.435  ns/op
+ArrayBufferBenchmark.addAll1   10000  avgt   30  25152.311 ± 136.803  ns/op
-ArrayBufferBenchmark.addAll2      10  avgt   30    106.333 ±   2.565  ns/op
+ArrayBufferBenchmark.addAll2      10  avgt   30    119.342 ±   0.267  ns/op
-ArrayBufferBenchmark.addAll2     100  avgt   30    669.213 ±   5.293  ns/op
+ArrayBufferBenchmark.addAll2     100  avgt   30    784.811 ±   9.628  ns/op
-ArrayBufferBenchmark.addAll2    1000  avgt   30   7041.905 ±  13.658  ns/op
+ArrayBufferBenchmark.addAll2    1000  avgt   30   8561.646 ±  14.924  ns/op
-ArrayBufferBenchmark.addAll2   10000  avgt   30  81940.910 ± 152.333  ns/op
+ArrayBufferBenchmark.addAll2   10000  avgt   30  92149.025 ± 294.481  ns/op
-ArrayBufferBenchmark.addAll3      10  avgt   30    115.073 ±   2.267  ns/op
+ArrayBufferBenchmark.addAll3      10  avgt   30    134.425 ±   0.757  ns/op
-ArrayBufferBenchmark.addAll3     100  avgt   30    770.600 ±   2.999  ns/op
+ArrayBufferBenchmark.addAll3     100  avgt   30    761.728 ±   1.517  ns/op
-ArrayBufferBenchmark.addAll3    1000  avgt   30   6750.484 ±  11.538  ns/op
+ArrayBufferBenchmark.addAll3    1000  avgt   30   6688.084 ±  15.008  ns/op
-ArrayBufferBenchmark.addAll3   10000  avgt   30  78802.133 ± 137.187  ns/op
+ArrayBufferBenchmark.addAll3   10000  avgt   30  76560.228 ± 181.348  ns/op
```

sure doesn't look like an optimisation 😬

NthPortal commented:
I re-ran the "after" benchmarks and got the same result.

I'm gonna remove the attempted optimisation and stick with just the cleanup

NthPortal commented:
the new benchmark results are... confusing

```diff
 Benchmark                     (size)  Mode  Cnt       Score     Error  Units
-ArrayBufferBenchmark.addAll1      10  avgt   30     100.307 ±   0.263  ns/op
+ArrayBufferBenchmark.addAll1      10  avgt   30      99.412 ±   0.221  ns/op
-ArrayBufferBenchmark.addAll1     100  avgt   30     267.558 ±   0.700  ns/op
+ArrayBufferBenchmark.addAll1     100  avgt   30     267.451 ±   0.619  ns/op
-ArrayBufferBenchmark.addAll1    1000  avgt   30    1711.113 ±  16.689  ns/op
+ArrayBufferBenchmark.addAll1    1000  avgt   30    1686.831 ±   1.987  ns/op
-ArrayBufferBenchmark.addAll1   10000  avgt   30   25082.094 ± 153.114  ns/op
+ArrayBufferBenchmark.addAll1   10000  avgt   30   24900.748 ±  38.106  ns/op
-ArrayBufferBenchmark.addAll2      10  avgt   30     117.481 ±   0.499  ns/op
+ArrayBufferBenchmark.addAll2      10  avgt   30     117.389 ±   2.800  ns/op
-ArrayBufferBenchmark.addAll2     100  avgt   30    1083.291 ±   4.457  ns/op
+ArrayBufferBenchmark.addAll2     100  avgt   30     901.295 ±  13.703  ns/op
-ArrayBufferBenchmark.addAll2    1000  avgt   30    7396.998 ±  19.380  ns/op
+ArrayBufferBenchmark.addAll2    1000  avgt   30    8899.769 ± 230.113  ns/op
-ArrayBufferBenchmark.addAll2   10000  avgt   30  116549.685 ± 243.404  ns/op
+ArrayBufferBenchmark.addAll2   10000  avgt   30  103605.259 ± 226.673  ns/op
-ArrayBufferBenchmark.addAll3      10  avgt   30     103.075 ±   0.366  ns/op
+ArrayBufferBenchmark.addAll3      10  avgt   30     109.616 ±   3.312  ns/op
-ArrayBufferBenchmark.addAll3     100  avgt   30     666.471 ±   5.343  ns/op
+ArrayBufferBenchmark.addAll3     100  avgt   30     638.888 ±   1.148  ns/op
-ArrayBufferBenchmark.addAll3    1000  avgt   30    7072.978 ±  13.349  ns/op
+ArrayBufferBenchmark.addAll3    1000  avgt   30    7053.111 ±  14.073  ns/op
-ArrayBufferBenchmark.addAll3   10000  avgt   30   81894.343 ± 131.434  ns/op
+ArrayBufferBenchmark.addAll3   10000  avgt   30   82046.005 ± 181.125  ns/op
-ArrayBufferBenchmark.addAll4      10  avgt   30     114.865 ±   2.215  ns/op
+ArrayBufferBenchmark.addAll4      10  avgt   30     121.173 ±   4.157  ns/op
-ArrayBufferBenchmark.addAll4     100  avgt   30     781.175 ±   4.043  ns/op
+ArrayBufferBenchmark.addAll4     100  avgt   30     747.862 ±   2.178  ns/op
-ArrayBufferBenchmark.addAll4    1000  avgt   30    6749.267 ±  15.244  ns/op
+ArrayBufferBenchmark.addAll4    1000  avgt   30    6750.256 ±  10.804  ns/op
-ArrayBufferBenchmark.addAll4   10000  avgt   30   79043.508 ± 275.626  ns/op
+ArrayBufferBenchmark.addAll4   10000  avgt   30   78989.568 ± 132.373  ns/op
```

`addAll2` is the one that the new changes are meant to optimise, and the results are highly inconsistent. sometimes moderate improvements, sometimes moderate losses.

Either `ArrayBuffer#addAll(s.c.i.Set)` has really weird behaviour, or something caused slowdowns during benchmark execution. I'm gonna rerun `addAll2` with a 20 warmup and 40 rounds (rather than 15 and 15), to hopefully get more consistent results.

NthPortal commented:
```diff
 Benchmark                     (size)  Mode  Cnt       Score     Error  Units
-ArrayBufferBenchmark.addAll2      10  avgt   80     120.388 ±   0.697  ns/op
+ArrayBufferBenchmark.addAll2      10  avgt   80     115.786 ±   0.649  ns/op
-ArrayBufferBenchmark.addAll2     100  avgt   80    1096.057 ±   2.217  ns/op
+ArrayBufferBenchmark.addAll2     100  avgt   80     941.184 ±   6.919  ns/op
-ArrayBufferBenchmark.addAll2    1000  avgt   80    7367.775 ±  16.810  ns/op
+ArrayBufferBenchmark.addAll2    1000  avgt   80    8626.615 ±  49.965  ns/op
-ArrayBufferBenchmark.addAll2   10000  avgt   80  103249.393 ± 517.055  ns/op
+ArrayBufferBenchmark.addAll2   10000  avgt   80  101780.067 ± 534.532  ns/op
```

... nope. I don't understand how, but nope. I'll revert the attempted optimisation completely, and just leave it at the tidying up and the `View.Patched` bugfix

SethTisue commented:
Performance is a cruel and distant god.

NthPortal committed:
Fix unreported bug in `View.Patched`

Fix unreported bug in `View.Patched` where iterator
is incorrect due to the patch only being iterable once,
and already having been exhausted.

Discovered while attempting to optimise `ArrayBuffer` in an earlier
version of the previous commit.

SethTisue commented:
2nd attempt to attract reviewer attention from @scala/collections crew and/or @julienrf

julienrf commented:
Hey, sorry I can’t review the PR soon.

joshlemer commented on src/library/scala/collection/View.scala:
```diff
@@ -404,9 +404,15 @@ object View extends IterableFactory[View] {
+    def iterator: Iterator[A] = underlying.iterator.patch(from, _other.iterator, replaced)
     override def knownSize: Int = if (underlying.knownSize == 0 && other.knownSize == 0) 0 else super.knownSize
-    override def isEmpty: Boolean = if (knownSize == 0) true else iterator.isEmpty
+    override def isEmpty: Boolean = if (knownSize == 0) true else _other.isEmpty
```
Does this logic work? Would this not be wrong for case 

```scala
List(1,2,3).view.patch(0, Seq(), 0).isEmpty
```
? Looks like with this logic the above would return `true` but in fact it is false.

NthPortal replied:
oops, good point. idk why I thought that was a good change. I'm blaming it on me from many months ago

NthPortal replied:
funnily enough, the previous line should use `_other` instead of `other` as well

johnynek commented on test/junit/scala/collection/mutable/ArrayBufferTest.scala:
```diff
@@ -455,4 +455,21 @@ class ArrayBufferTest {
+  @Test
+  def viewConsistency(): Unit = {
+    def check[U](op: ArrayBuffer[Int] => U): Unit = {
+      val buf = ArrayBuffer(1, 2, 3)
```
could we instead use scalacheck here? It's not clear there aren't edge cases at empty, single element, etc...
```scala
forAll(genArrayBuffer) { buf =>
 ...
}
```

NthPortal replied:
what would the code after that be? sorry, I'm not super familiar with scalacheck, and I'm not following your suggestion

johnynek replied:
```scala
forAll { (ints: List[Int]) =>

  def check(op: ArrayBuffer[Int] => Any): Unit = {
    // make a new buf each time
    val buf = ArrayBuffer(ints: _*)
    val view = buf.view
    op(buf)
    assertSameElements(buf, view)
  }

  check(_.clear())
  check(_dropRightInPlace(1))
  check(_.dropInPlace(1))
  check(_ ++= (1 to 100))
  check (_.insertAll(1, 1 to 100)) 
}
```

IME scalacheck is a massive win on bug finding. I virtually always find bugs when I add scalacheck tests.

NthPortal replied:
ftr, this is missing a bunch of stuff, including:

- `assertSameElements` should just be a call to `sameElements`, so that it's a `Boolean` rather than something that throws
- the calls to `check` need to be combined (using `&&`? `++`?) so that all but the last aren't ignored
- `insertAll(1, ???)` will throw an exception if the `ArrayBuffer` is empty
- `++=` and `insertAll` may throw an exception if the `ArrayBuffer` is larger than `Int.MaxValue / 2 - 100` (that value is based on the implementation, and theoretically should allow larger buffers)

relating to that second point, I was looking at some other classes to try to understand how scalacheck works, and it appears that there are other places that drop checks on the floor, including https://github.com/scala/scala/blob/66e8d34e646a616f39615cbf7e8fc6d56103b2ec/test/scalacheck/scala/collection/immutable/VectorMapProperties.scala#L38-L42

NthPortal replied:
I've resolved this (added scalacheck property tests), but I'm gonna leave it unresolved for now as a reminder that some other property tests are broken

NthPortal replied:
regarding `VectorMapProperties`: see #9768

johnynek commented on test/junit/scala/collection/ViewTest.scala:
```diff
@@ -113,4 +113,11 @@ class ViewTest {
+  @Test
+  def patch(): Unit = {
+    val view = List(2).view.patch(1, List(3, 4, 5).iterator, 0)
+    assertEquals(view.toList, view.toList)
```
can we also assert what the value should be?

Also, could we test a patch law, something like:

```scala
forAll { (sq: SeqView[Int], from: Int, items: IterableOnce[Int], r: Int) =>
  val patchedView = Try(sq.patch(from, items, r)).toOption

  val viaList = Try {
    val (left, right) = sq.toList.splitAt(from)
    (left ++ items ++ right.drop(r)).toList
  }.toOption
 
  val asList = patchedView.map(_.toList)
  (asList == viaList) &&
    // make sure the view doesn't change
    (patchedView.map(_.toList) == viaList) 
}
```

NthPortal replied:
that code doesn't work, because `items` cannot be reused. I've added an assertion as to the actual value, however

johnynek replied:
yes, sorry. Make `items` a `List[Int]` and then pass a `.iterator` of it as needed.

NthPortal replied:
`patch` actually doesn't care if `from` is past the end of the original collection, so `Try` wasn't needed. also, using `Try#toOption` would have likely:

- not worked because if exactly one of the `Try`s fails, the whole things fails with no information.
- hidden the bug that cropped up.

also, there's no `Arbitrary` for `SeqView`, and you end up needing to use specifically a `Seq` that doesn't extend `StrictOptimisedSeqOps` to have a meaningful test. hopefully it doesn't get changed in a future version and make the test worthless.

NthPortal replied:
regardless, I've tweaked it to work

johnynek reviewed (commented):
maybe I suffer from anxiety, but given this is in the standard library, and given that we already have found bugs, I think we need more extensive tests. I've suggested two scalachecks to add.

SethTisue commented:
@NthPortal what's next here? are you interested in doing the ScalaCheck thing @johnynek suggested?

NthPortal commented:
@SethTisue i've been and continue to be sick, but i'll try to add the scalacheck stuff today or tomorrow. if i don't get to it, and @johnynek has perms, they can feel free to add a commit to this branch with it

NthPortal committed:
Fix unreported bug in `ArrayOps.ArrayIterator`

Fix unreported bug in `ArrayOps.ArrayIterator` where
index in array can overflow to negative, and in such cases
the iterator reports `hasNext` incorrectly as `true`.

Improve code documentation and readability of
`Iterator.patch` implementation, which was bumped
into while investigating the bug.

Discovered while writing tests for bug in `View.Patched`
(see following commit).

NthPortal commented:
oops I've bumped into ANOTHER unreported bug

NthPortal commented:
this is done as far as I'm concerned

NthPortal commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -317,18 +312,30 @@ object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
+    }, () => 0)
   }
 
+  @deprecated("never intended to be public", since = "2.13.6")
```
this PR has been open too long 😅

```suggestion
  @deprecated("never intended to be public", since = "2.13.7")
```

NthPortal commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -317,18 +312,30 @@ object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
+// TODO: use `CheckedIndexedSeqView.Id` once we can change the return type of `ArrayBuffer#view`
+final class ArrayBufferView[A] private[mutable](underlying: ArrayBuffer[A], mutationCount: () => Int)
   extends AbstractIndexedSeqView[A] {
   @deprecated("never intended to be public; call ArrayBuffer#view instead", since = "2.13.6")
```
```suggestion
  @deprecated("never intended to be public; call ArrayBuffer#view instead", since = "2.13.7")
```
